lib/db/mysql/index.js

```diff
@@ -606,38 +614,159 @@ MysqlStore.prototype = {
             return;
           }
 
-          var clientIn = ignoreClientId.map(function(id) {
+          const clientIn = ignoreClientId.map((id) => {
             return buf(id);
           });
 
           return self._write(QUERY_PURGE_EXPIRED_TOKENS, [clientIn, deleteBatchSize])
-            .then(function (res) {
+            .then((res) => {
               logger.info('purgeExpiredTokens', { affectedRows: res.affectedRows });
 
               // Break loop if no items were effected by delete.
               // All expired tokens have been deleted.
               if (res.affectedRows === 0) {
-                message = '0 affectedRows. Bailing out.';
+                const message = '0 affectedRows. Bailing out.';
                 logger.info('purgeExpiredTokens', { message: message });
                 return;
               }
 
               deletedItems = deletedItems + res.affectedRows;
 
               return P.delay(delaySeconds * 1000)
-                .then(function () {
+                .then(() => {
                   return promiseWhile();
                 });
             });
         });
 
         return promiseWhile();
       })
-      .then(function() {
+      .then(() => {
         logger.info('purgeExpiredTokens', { message: 'completed' });
       });
   },
 
+
+  // This version of purgeExpiredTokens uses the strategy of selecting a set
+  // of tokens to delete and then issuing deletes by primary key.
+  purgeExpiredTokensById: function purgeExpiredTokensById(numberOfTokens,
+                                                          delaySeconds,
+                                                          ignoreClientId,
+                                                          deleteBatchSize = 200)
+  {
+    const self = this;
+    if (! ignoreClientId) {
+      throw new Error('empty ignoreClientId');
+    }
+
+    if (! Array.isArray(ignoreClientId)) {
+      ignoreClientId = [ ignoreClientId ];
+    }
+
+    if (numberOfTokens <= deleteBatchSize) {
+      deleteBatchSize = numberOfTokens;
+    }
+
+    const clientIds = ignoreClientId.map((id) => {
+      return self.getClient(id);
+    });
+
+    let lastPurgeTime;
+
+    return P.all(clientIds)
+      .then((results) => {
+        // This ensures that purgeExpiredTokensById can not be called with an
+        // unknown ignoreClientId(s).
+        results.forEach((ignoreClient) => {
+          if (! ignoreClient) {
+            throw new Error('unknown ignoreClientId ' + ignoreClientId);
+          }
+        });
+      })
+      .then(() => {
+        // Continue from the last recorded 'last-purge-time', if available.
+        return self._write(QUERY_LAST_PURGE_TIME)
```

Comment:
nit: this is a read, should we use `self._read` for this?  They currently do the same thing (just calling `this._query` but they may not in the future.

Reply:
Actually `self._readOne` would be even better here and would avoid the need to defer `res[0]`.

Reply:
Updated to use readOne, or read where I want a list. Thanks.